Full Airflow Integration (With Foreachs)

valayDave opened the pull request.

valayDave committed:
commit 0
- unfinished code.

valayDave committed:
more tweeks.

valayDave committed:
First version of dynamic dag compiler.
- Not completely finished code
- Creates generic .py file a JSON that is parsed to create Airflow DAG.
- Currently only boiler plate to make a linear dag but doesn't execute anything.
-  Unfinished code.

valayDave committed:
some changes based on savin's comments.
- Added changes to task datastore for different reason : (todo) Decouple these
- Added comments to SFN for reference.
- Airflow DAG is no longer dependent on metaflow

valayDave committed:
Tweeks

valayDave committed:
Got linear workflows working on airflow.
- Still not feature complete as lots of args are still unfilled / lots of unknows
- minor tweek in eks to ensure airflow is k8s compatible.
- passing state around via xcom-push
- HACK : AWS keys are passed in a shady way. : Reverse this soon.

valayDave committed:
Fixed setting run-id / task-id to labels in k8s
- Fixed setting run-id has from cli macro
- added hashing macro to ensure that jinja template set the correct run-id to k8s labels
-

valayDave committed:
Added @schedule support + readme

valayDave committed:
Nit fix.

valayDave committed:
setting retry logic.

valayDave committed:
Fixed setting `task_id` :
- switch task-id from airflow job is to hash to "runid/stepname"
- refactor xcom setting variables
- added comments

valayDave committed:
Changed retry setting.

valayDave committed:
Fixed bug when using catch + retry

valayDave committed:
making changes sync with master

valayDave committed:
Merge branch 'master' into aft-mm

valayDave committed:
Fixing logs related change from master.

valayDave committed:
Removed todos + added some validation logic.

valayDave committed:
Added support for Parameters.
- Supporting int, str, bool, float, JSONType

valayDave committed:
Removed hacky line.

valayDave committed:
Added hacky line

valayDave committed:
Added support for Branching with Airflow
- remove `next` function in `AirflowTask`
- `AirflowTask`s has no knowledge of next tasks.
- removed todos + added some todos
- Graph construction on airflow side using graph_structure datastructure.
- graph_structure comes from`FlowGraph.output_steps()[1]`

valayDave committed:
fixing bugs in service account setting

valayDave committed:
2 changes :
- removed hacky line
- added support to directly throw dags in s3.

valayDave committed:
Adding exceptions for missing features.

valayDave committed:
Max workers and worker pool support.

valayDave committed:
@project, @schedule, default active dag support.
- Added a flag to allow setting dag as active on creation
- Airflow compatible schedule interval
- Project name fixes.

valayDave committed:
Bug fix in active defaults.

valayDave committed:
bug fix.

valayDave committed:
Bug fix.

valayDave committed:
Added secrets support.

valayDave committed:
Bug fixes WRT Kubernetes secrets + k8s deployments.
- Fixing some error messages.
- Added some comments.

valayDave committed:
Fixing bug fix in schedule.

valayDave committed:
Removing un-used code paths + code cleanup

valayDave committed:
Changes based on PR comments:
- removed airflow xcom push file , moved to decorator code
- removed prefix configuration
- nit fixes.

valayDave committed:
Moved exceptions to airflow compiler

valayDave committed:
Removing un-used args.

valayDave committed:
shortened length of run-id and task-id hashes.

valayDave committed:
changed `--is-active`->`--is-paused-upon-creation`
- dags are active by default.

valayDave committed:
Made file path required arguement.

valayDave committed:
Remove readme.

valayDave committed:
Removed usless code WRT project decorator.

valayDave committed:
Fixing mf log related bug
- No double logging on metaflow.

valayDave committed:
Added validations about:
- un-supported decorators
- foreach
Changed where validations are done to not save the package.

valayDave committed:
Refactored code .
- removed compute/k8s.py
- Moved k8s code to airflow_compiler.py
- ran isort to airflow_compiler.py

valayDave committed:
Fixed passing secrets with kubernetes.

valayDave committed:
3 changes:
- Removing s3 dep
- remove uesless import
- added `deployed_on` in dag file template

valayDave committed:
reverting change.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow-tests

valayDave committed:
ran black,

valayDave committed:
Removing un-used code / resolved-todos.

valayDave committed:
Merge branch 'airflow-tests' into airflow

valayDave committed:
variable renamemetaflow_specific_args

valayDave committed:
Option help bug fix.

valayDave committed:
fixed printing bug

valayDave committed:
Addressed comments.

valayDave committed:
Airflow sensor api (#3)

* Fixed run-id setting
- Change gaurentees that multiple dags triggered at same moment have unique run-id

* added allow multiple in `Decorator` class

* Airflow sensor integration.
 >> support added for :
- ExternalTaskSensor
- S3KeySensor
- SqlSensor
>> sensors allow multiple decorators
>> sensors accept those arguments which are supported by airflow

* Added `@airflow_schedule_interval` decorator
* Fixing bug run-id related in env variable setting.

valayDave committed:
porting savin's comments
- next changes : addressing comments.

valayDave committed:
Removed Sensor metadata extraction.

valayDave committed:
Multiple Changes based on comments:
1. refactored `create_k8s_args` into _to_job
2. Addressed comments for snake casing
3. refactored `attrs` for simplicity.
4. refactored `metaflow_parameters` to `parameters`.
5. Refactored setting of `input_paths`

valayDave committed:
Following Changes:
- Refactors setting scheduling interval
- refactor dag file creating function
- refactored is_active to is_paused_upon_creation
- removed catchup

valayDave committed:
Fixed defaults passing
- Addressed comments for airflow.py

valayDave committed:
nit fixes to dag.py's templating variables.

valayDave committed:
nit fix : variable name change.

valayDave committed:
added mising Run-id prefixes to variables.
- merged `hash` and `dash_connect` filters.

valayDave committed:
Merge branch 'master' into af-mmr

valayDave committed:
squashing bugs after changes from master.

valayDave committed:
Moving information from airflow_utils to compiler (#56)

- commenting todos to organize unfinished changes.
- some environment variables set via`V1EnvVar`
    - `client.V1ObjectFieldSelector` mapped env vars were not working in json form
    - Moving k8s operator import into its own function.
    - env vars moved.

valayDave committed:
ran black.

valayDave committed:
Refactored sensors to different files + bug fix
- bug caused due `util.compress_list`.
- The function doesn't play nice with strings with variety of characters.
- Ensured that exceptions are handled appropriately.
- Made new file for each sensor under `airflow.sensors` module.

valayDave committed:
GPU support added + container naming same as argo

valayDave committed:
Removed unused `sanitize_label` function

valayDave committed:
removed un-used function

valayDave committed:
k8s bug fix

valayDave committed:
Addressing Final comments.  (#57)

- Added dag-run timeout.
- airflow related scheduling checks in decorator.
- Auto naming sensors if no name is provided
- Annotations to k8s operators
- fix: argument serialization for `DAG` arguments (method names refactored like `to_dict` became `serialize`)
- annotation bug fix
- setting`workflow-timeout` for only scheduled dags

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Bug fix from master merge.

valayDave committed:
addressing  savins comments. (#59)

- Added many adhoc changes based for some comments.
- Integrated secrets and `KUBERNETES_SECRETS`
- cleaned up parameter setting
- cleaned up setting of scheduling interval
- renamed `AIRFLOW_TASK_ID_TEMPLATE_VALUE` to `AIRFLOW_TASK_ID`
- renamed `AirflowSensorDecorator.compile` to `AirflowSensorDecorator.validate`
- Checking if dagfile and flow file are same.
- fixing variable names.
- checking out `kubernetes_decorator.py` from master (6441ed5)
- bug fixing secret setting in airflow.
- simplified parameter type parsing logic
- refactoring airflow argument parsing code.

valayDave committed:
refactored an error string

valayDave committed:
refactor :
-`RUN_ID_LEN` to `RUN_HASH_ID_LEN`
- `TASK_ID_LEN` to `TASK_ID_HASH_LEN`

valayDave committed:
Refactored an `id_creator` method.

valayDave committed:
added comment on need for `start_date`

valayDave committed:
Refactored parameter macro settings. (valayDave/metaflow#60)

valayDave committed:
simplifying run-id macro.

valayDave committed:
Airflow Foreach Integration
- Simple one node foreach-join support as gaurenteed by airflow
- Fixed env variable setting issue
- introduced MetaflowKuberentesOperator
- Created a new operator to allow smootness in plumbing xcom values
- Some todos

valayDave committed:
Adding default 1 retry for any airflow worker.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
updated pod startup time

valayDave committed:
Removed retries from user-defaults.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Foreach polish (valayDave/metaflow#62)


* Removing unused imports
* Added validation logic for airflow version numbers with foreaches
* Removed `airflow_schedule_interval` decorator.

* Added production/deployment token related changes
- Uses s3 as a backend to store the production token
- Token used for avoiding nameclashes
- token stored via `FlowDatastore`

* Graph type validation for airflow foreachs
- Airflow foreachs only support single node fanout.
- validation invalidates graphs with nested foreachs

* Added configuration about startup_timeout.

* Added final todo on `resources` argument of k8sOp
- added a commented code block
- it needs to be uncommented when airflow releasese the patch for the op
- Code seems feature complete keeping aside airflow patch

savingoyal commented on metaflow/decorators.py:
```diff
@@ -469,11 +473,27 @@ def _attach_decorators_to_step(step, decospecs):
     flow, graph, environment, flow_datastore, metadata, logger, echo, deco_options
 ):
+    # Since _flow_decorators is a dictonary and some flow-decorators can have many decos of same type allowed.
+    # Hence we set certain keys as lists to accomodate multiple decos of same type and therefore validate if the value is a list
```
Minor nit - `Certain decorators can be specified multiple times and exist as lists in the  _flow_decorators dictionary`

savingoyal commented on metaflow/decorators.py:
```diff
@@ -469,11 +473,27 @@ def _attach_decorators_to_step(step, decospecs):
-            flow, graph, environment, flow_datastore, metadata, logger, echo, opts
-        )
+        if type(deco) == list:
+            for rd in deco:
```
You can actually simplify this entire logic by having a dictionary that has only lists as values.

valayDave replied:
Many underlying constructs expect `_flow_decorators` of the form `key:value` over `key:[values]`. So making this change would require me refactoring those other places too. I felt that to change airflow I didn't want to alter other parts of the code so I did it this way.

savingoyal commented on metaflow/metaflow_config.py:
```diff
@@ -240,6 +240,14 @@ def from_conf(name, default=None):
+##
+# Airflow Configuration
+##
+AIRFLOW_KUBERNETES_STARTUP_TIMEOUT = from_conf(
```
Can you add a comment here notating where the configuration is actually used?

valayDave replied:
Done.

savingoyal commented on metaflow/plugins/airflow/exception.py:
```diff
@@ -0,0 +1,12 @@
+from metaflow.exception import MetaflowException
```
exceptions.py

valayDave replied:
You mean change file name ?

savingoyal commented on metaflow/plugins/__init__.py:
```diff
@@ -161,7 +165,18 @@ def get_plugin_cli():
 
-FLOW_DECORATORS = [CondaFlowDecorator, ScheduleDecorator, ProjectDecorator]
+
+from .airflow.sensors import (
```
can you just import `SUPPORTED_SENSORS` here from `sensors/_init_.py` (`SUPPORTED_SENSORS` would need some modification for it to work properly though - classes instead of names).

savingoyal commented on metaflow/plugins/airflow/sensors/base_sensor.py:
```diff
@@ -0,0 +1,74 @@
+
+    allow_multiple = True
+
+    defaults = dict(
```
Is there an Airflow object that can be imported instead? Otherwise, we will be on the hook for maintaining this list by ourselves, and keeping it backwards compatible is going to be a big issue.

savingoyal replied:
This is used on the user's workstation so harder to keep it dependent on Airflow. Ignore the above comment.

savingoyal commented on metaflow/plugins/airflow/sensors/base_sensor.py:
```diff
@@ -0,0 +1,74 @@
+            operator_type=self.operator_type,
+        ).set_operator_args(**{k: v for k, v in task_args.items() if v is not None})
+
+    def validate(self):
```
There is no actual validation happening here - we can perhaps rename this to `santize`

valayDave replied:
This function is inherited and called by subclasses which do the validation. Check any sensor using the base sensor.

savingoyal commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -127,7 +127,6 @@ def step_init(self, flow, graph, step, decos, environment, flow_datastore, logge
             raise KubernetesException(
                 "The *@kubernetes* decorator requires --datastore=s3 at the moment."
             )
-
```
This file can be removed from the PR.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+                (
+                    cls.get_token_path(token),
+                    BytesIO(
+                        bytes(json.dumps({"token": token, "owner": owner}), "utf-8")
```
rename `token` to `production_token`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            return "@daily"
+        return None
+
+    def _get_airflow_schedule_interval(self):
```
Didn't we drop `@airflow_schedule_interval`?

valayDave replied:
Fair. Forgot to remove it from there. Will change.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+    def _get_retries(self, node):
+        max_user_code_retries = 0
+        max_error_retries = 0
+        foreach_default_retry = 1
```
Why do we have a default retry for `foreach`?

valayDave replied:
Because foreach's are flaky with Airflow. As cardinality grows the chance of failure grows by power law.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+                raise MetaflowException(
+                    "Parameter *%s* does not have a "
+                    "default value. "
+                    "A default value is required for parameters when deploying on Airflow."
```
`when deploying flows to Airflow.`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            if value is not None:
+                airflow_param["default"] = value
+            if param_help:
+                airflow_param["description"] = param_help
```
It can be a simple `airflow_param["description"] = param.kwargs.get("help")`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            )
+            param_help = param.kwargs.get("help")
+            if value is not None:
+                airflow_param["default"] = value
```
`airflow_param["default"] = deploy_time_eval(param.kwargs.get("default"))`

valayDave replied:
This is already done in line number 204

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+
+        return airflow_params
+
+    def _make_input_path_compressed(
```
`def _compress_input_path`

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            self._make_input_path(s, only_task_id=True) for s in step_names
+        )
+
+    def _make_foreach_input_path(self, step_name):
```
Does this need to be a function?

valayDave replied:
Yes because there is a macro used inside it which is evaluated at DAG runtime to join the values as a comma-separated list.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+
+    def _make_input_path_compressed(
+        self,
+        step_names,
```
steps

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+        Hence we have a custom compression string created via `_make_input_path_compressed` function instead of `compress_list`.
+        """
+        return "%s:" % (AIRFLOW_MACROS.RUN_ID) + ",".join(
+            self._make_input_path(s, only_task_id=True) for s in step_names
```
for step in steps

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+        )
+
+    def _make_input_path(self, step_name, only_task_id=False):
+        # This is set using the `airflow_internal` decorator.
```
Can you compress the 4 comments in one?

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            1. Based on the type of the graph node (start/linear/foreach/join etc.) it will decide how to set the input paths
+            2. Based on node's decorator specification convert the information into a job spec for the KubernetesPodOperator.
+        """
+        # supported compute : k8s (v1), local(v2), batch(v3)
```
You can remove this comment

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+        if env_deco:
+            env = env_deco[0].attributes["vars"]
+
+        is_foreach_join = (
```
You can define this closer to where it is used to aid in readability.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            node.type == "join" and self.graph[node.split_parents[-1]].type == "foreach"
+        )
+
+        # The Below If/Else Block handle "Input Paths".
```
Fix capitalization for comments.

savingoyal commented on metaflow/plugins/airflow/airflow.py:
```diff
@@ -0,0 +1,727 @@
+            # parameters.
+
+            if len(self.parameters):
+                env["METAFLOW_PARAMETERS"] = AIRFLOW_MACROS.PARAMETER
```
nit - AIRFLOW_MACROS.PARAMETERS

savingoyal reviewed (commented):
Still reviewing

valayDave committed:
Removing traces of `@airflow_schedule_interval`

valayDave committed:
new comment on `startup_timeout_seconds` env var.

valayDave committed:
Refactored import Airflow Sensors.

valayDave committed:
change `token` to `production_token`

valayDave committed:
nit fixes :
- fixing comments.
- refactor some variable/function names.

valayDave committed:
Merge branch 'master' into airflow

valayDave committed:
Uncommented code for foreach support with k8s

KubernetesPodOperator version 4.2.0 renamed `resources` to
`container_resources`
- Check : (https://github.com/apache/airflow/pull/24673) /
- (https://github.com/apache/airflow/commit/45f4290712f5f779e57034f81dbaab5d77d5de85)

This was done because `KubernetesPodOperator` didn't play nice with dynamic task mapping and they had to deprecate the `resources` argument. Hence the below codepath checks for the version of `KubernetesPodOperator`
and then sets the argument. If the version < 4.2.0 then we set the argument as `resources`.
If it is > 4.2.0 then we set the argument as `container_resources`
The `resources` argument of KuberentesPodOperator is going to be deprecated soon in the future.
So we will only use it for `KuberentesPodOperator` version < 4.2.0
The `resources` argument will also not work for foreach's.

valayDave committed:
commiting k8s related file from master.

valayDave committed:
Merge branch 'master' into airflow
